source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2515 @@
+  information.
+
+  The array should contain at least one document. The structure of each document
+  is as follows:
+
+  - ``client``: Required string. Client entity on which the events are expected
+    to be observed. See `commonOptions_client`_.
+
+  - ``events``: Required array of documents. List of events, which are expected
+    to be observed (in this order) on the corresponding client while executing
+    `operations`_. If the array is empty, the test runner MUST assert that no
+    events were observed on the client (excluding ignored events).
+
+    The structure of each document is defined in `expectedEvent`_.
+
+.. _test_outcome:
+
+- ``outcome``: Optional array of documents. Each document will specify expected
+  contents of a collection after all operations have been executed. The list of
+  documents therein SHOULD be sorted ascendingly by the ``_id`` field to allow
+  for deterministic comparisons.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `collectionData`_.
+
+
+operation
+~~~~~~~~~
+
+An operation to be executed as part of the test.
+
+The structure of this document is as follows:
+
+.. _operation_name:
+
+- ``name``: Required string. Name of the operation (e.g. method) to perform on
+  the object.
+
+.. _operation_object:
+
+- ``object``: Required string. Name of the object on which to perform the
+  operation. This should correspond to either an `entity`_ name (for
+  `Entity Test Operations`_) or "testRunner" (for `Special Test Operations`_).
+  If the object is an entity, The YAML file SHOULD use an `alias node`_ for its
+  ``id`` field (e.g. ``object: *collection0``).
+
+.. _operation_arguments:
+
+- ``arguments``: Optional document. Map of parameter names and values for the
+  operation. The structure of this document will vary based on the operation.
+  See `Entity Test Operations`_ and `Special Test Operations`_.
+
+  The ``session`` parameter is handled specially (see `commonOptions_session`_).
+
+.. _operation_expectedError:
+
+- ``expectedError``: Optional document. One or more assertions for an expected
+  error raised by the operation. The structure of this document is
+  defined in `expectedError`_.
+
+  This field is mutually exclusive with
+  `expectedResult <operation_expectedResult_>`_ and
+  `saveResultAsEntity <operation_saveResultAsEntity_>`_.
+
+.. _operation_expectedResult:
+
+- ``expectedResult``: Optional mixed type. A value corresponding to the expected
+  result of the operation. This field may be a scalar value, a single document,
+  or an array of documents in the case of a multi-document read.  Test runners
+  MUST follow the rules in `Evaluating Matches`_ when processing this assertion.
+
+  This field is mutually exclusive with
+  `expectedError <operation_expectedError_>`_.
+
+.. _operation_saveResultAsEntity:
+
+- ``saveResultAsEntity``: Optional string. If specified, the actual result
```

Comment:
I continue to think about unifying `entities` and `operations`, and the interesting cross-section.

```
{
  reference: <unique-id-of-method-function-class>, // op has this as name, entity has this as parent key,
  entityId: '' // operation has this as `saveResultAsEntity`, entity has this as id. 
  meta: {
    
  }
  parameters: { // arguments / parameters

  }
  expectSuccess: { // can be absent for entities; 

  }
  expectFailure: { // can be absent for entities;

  }
}
```

Reply:
Per my earlier comment, not every entity definition directly corresponds to an API method. `MongoClient.startSession` may be the only case where we do have a specified way to construct an entity.

I'm open to renaming `expectedResult` and `expectedError` to `expectResult` and `expectFailure`, though, as the language is both shorter and more active. I don't prefer `expectSuccess`, since the value of `expectedResult` is a match on the return value -- and `expectSuccess` sounds like it'd more generally be used to assert that an operation did not raise an error.

Also not a fan of `entityId` being used as a common syntax for `saveResultAsEntity`, as it drops the active language. `saveResultAsEntity` very clearly communicates what is happening.

Regarding `arguments` and `parameters`, I propose we take a poll during final review and see what is most preferred by the team.

Reply:
I'd slightly prefer `arguments` to `parameters` as it's more in line with the technical definitions of those terms (i.e. parameters define what should be passed in, arguments are what you actually pass in). I don't have a strong preference about `expected` vs. `expect`, but I think we should make sure `expectedEvents` is consistent with whatever we land on.

Reply:
Just renamed the field names for `expectResult`, `expectError`, and `expectEvents`. I left the structures (e.g. `expectedEvent`, `expectedEventsForClient`, and `expectedError`) as-is, since those are only referred to in the spec itself and are never written in a test file; a described noun made more sense than an active phrase in those cases.